### ISSUE TITLE
Lab 5.2 Messed up output

### ISSUE THREAD

**Issue by zzhangyue** [nus-cs2030/2223-s2#724]
# Summary

It appears that my output is not in sequence and I just cannot see why...As a result, I cannot tell if my self-checks are working correctly too...Would appreciate some help here :< Thanks so much :(

<img width="598" alt="Screenshot 2023-04-03 at 20 41 12" src="https://user-images.githubusercontent.com/122248591/229512861-67b1c260-10a5-4c7a-9d97-d07380a22161.png">
<img width="596" alt="Screenshot 2023-04-03 at 20 41 33" src="https://user-images.githubusercontent.com/122248591/229512872-82b64187-d87e-4003-9f40-a8dbe8e6e86d.png">

**Comment by zzhangyue** [comment]
To add on, because it is possible that a customer can move around to any self-check while waiting before actually getting served by one self-check, can I get some advice on how I can constantly keep track of the difference in waiting time (get shortest waiting time) every time a self-check is updated. Like instead of waiting using server's next free time, it is possible for a customer to move to a self-check before this next free time, which linked to a intermeditateWait/Wait event...(if that makes sense...) Thanks!

**Comment by prisciliapangg** [comment]
I would think that there is nothing wrong with your polling and stuff (the screenshot). However, I think the problem lies in your comparator. I am assuming that you used .getTime() in your Comparator to determine which is to be polled out first. Hence, the problem is probably in the .getTime() of the respective events.

**Comment by prisciliapangg** [comment]
> To add on, because it is possible that a customer can move around to any self-check while waiting before actually getting served by one self-check, can I get some advice on how I can constantly keep track of the difference in waiting time (get shortest waiting time) every time a self-check is updated. Like instead of waiting using server's next free time, it is possible for a customer to move to a self-check before this next free time, right? Which linked to a IntermeditateWait/Wait event...(if that makes sense...) Thanks!

I don't think it is possible for a customer to move to a self-check (to be served) before the next free time. I found the earliest next time free by iterating through all the SCOs and finding the minimum, and then passing it into the IntermedaiteWait/Wait event, so that it can be polled out at the correct time.

**Comment by 0liverkong** [comment]
the fact that you have to wrap your data within 3 nested pairs might give you a hint that maybe a class to encapsulate the state of the shop may help.... just something to ponder if you have time.

**Comment by granolaluvr69** [comment]
> To add on, because it is possible that a customer can move around to any self-check while waiting before actually getting served by one self-check, can I get some advice on how I can constantly keep track of the difference in waiting time (get shortest waiting time) every time a self-check is updated. Like instead of waiting using server's next free time, it is possible for a customer to move to a self-check before this next free time, right? Which linked to a IntermeditateWait/Wait event...(if that makes sense...) Thanks!

I think that to do this, when creating an intermediate or a serve event (in wait / another intermediate event), you should have a method that iterates through all the next time avail for all the self checkouts, and sets the time of the new intermediate or serve event to this next avail timing. 

You should also ensure that your most updated server list is passed into every event so that the timeavail is always kept up to date.

Doing both of these together will ensure that the creation of the serve event will always be based on the most up to date timeavail of the selfcheckouts. 

Hope I helped!

**Comment by aowz** [comment]
The issue doesn't seem to lie with your polling methods. Seems to be with the event comparator itself. Try to prioritise by time then customerID. Secondly, you can also check how you pass a new time to your serve event from wait and whether that matches with your ToString method. Hope this helps
